// react-tables/dist-modules/TableBody.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; desc = parent = getter = undefined; _again = false; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) subClass.__proto__ = superClass; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var TableBody = (function (_React$Component) {
  function TableBody(props, context) {
    _classCallCheck(this, TableBody);

    _get(Object.getPrototypeOf(TableBody.prototype), 'constructor', this).call(this, props, context);

    this.getTr = this.getTr.bind(this);
  }

  _inherits(TableBody, _React$Component);

  _createClass(TableBody, [{
    key: 'getTr',
    value: function getTr(row, index) {
      var _context = this.context;
      var columns = _context.columns;
      var rows = _context.rows;

      if (!columns) return null;

      return _react2['default'].createElement(
        'tr',
        { key: index },
        columns.map(function (_ref) {
          var Component = _ref.Component;
          var formatter = _ref.formatter;
          var field = _ref.field;
          var title = _ref.title;
          var classes = _ref.classes;

          var style = {};
          if (!classes) {
            style.width = parseFloat(100 / columns.length).toFixed(2) + '%';
          }

          if (Component) {
            return _react2['default'].createElement(
              'td',
              { key: title, className: classes || '', style: style },
              ' ',
              _react2['default'].createElement(Component, { row: row }),
              ' '
            );
          } else if (formatter) {
            return _react2['default'].createElement('td', { key: title, className: classes || '', style: style, dangerouslySetInnerHTML: { __html: formatter(row[field], row) } });
          } else {
            return _react2['default'].createElement(
              'td',
              { key: title, className: classes || '', style: style },
              row[field] || '-'
            );
          }
        })
      );
    }
  }, {
    key: 'render',
    value: function render() {
      var _context2 = this.context;
      var columns = _context2.columns;
      var loading = _context2.loading;
      var rows = _context2.rows;
      var classes = _context2.classes;
      var dimentions = _context2.dimentions;

      return _react2['default'].createElement(
        'div',
        { style: { height: dimentions.height, overflow: 'auto' } },
        _react2['default'].createElement(
          'table',
          { className: classes },
          _react2['default'].createElement(
            'tbody',
            null,
            loading ? _react2['default'].createElement(
              'tr',
              null,
              _react2['default'].createElement(
                'td',
                { colSpan: columns.length, style: { textAlign: 'center' } },
                'Loading...'
              )
            ) : rows.map(this.getTr),
            loading || rows.length ? null : _react2['default'].createElement(
              'tr',
              null,
              _react2['default'].createElement(
                'td',
                { colSpan: columns.length, style: { textAlign: 'center' } },
                'No information'
              )
            )
          )
        )
      );
    }
  }], [{
    key: 'contextTypes',
    value: {
      columns: _react2['default'].PropTypes.array,
      loading: _react2['default'].PropTypes.bool,
      rows: _react2['default'].PropTypes.array,
      classes: _react2['default'].PropTypes.string,
      dimentions: _react2['default'].PropTypes.object
    },
    enumerable: true
  }]);

  return TableBody;
})(_react2['default'].Component);

exports['default'] = TableBody;
module.exports = exports['default'];
